host/sgx/linux/vdso.c

```diff
@@ -0,0 +1,239 @@
+            return_args->rdi = arg1;
+            return_args->rsi = arg2;
+            result = 0;
+            break;
+        }
+        case ENCLU_ERESUME:
+        {
+            /* Hardware exceptions occur */
+
+            oe_host_exception_context_t host_context = {0};
+
+            host_context.rax = ENCLU_ERESUME;
+            host_context.rbx = run->tcs;
+
+            /* AEP is assigned by vDSO implementation */
+
+            OE_TRACE_INFO("vDSO: exception occurred");
+
+            oe_host_handle_exception(&host_context);
+
+            result = ENCLU_ERESUME;
+            break;
+        }
+    }
+
+done:
+    /* If the result <= 0, the value will be forwared as the return
+     * value of _vdso_sgx_enter_enclave. Otherwise, _vdso_sgx_enter_enclave
+     * will invoke the ENCLU[result] instead of returning to the caller. */
+    return result;
+}
+
+/* The function should never be inline to preserve the stack frame. */
+OE_NEVER_INLINE
+oe_result_t oe_vdso_enter(
+    void* tcs,
+    uint64_t arg1,
+    uint64_t arg2,
+    uint64_t* arg3,
+    uint64_t* arg4,
+    oe_enclave_t* enclave)
+{
+    oe_ecall_context_t ecall_context = {0};
+    oe_result_t result = OE_UNEXPECTED;
+    struct sgx_enclave_run run = {0};
+    oe_vdso_return_args_t return_args = {0};
+    int return_value = 0;
+    uint32_t mxcsr = 0;
+    uint16_t fcw = 0;
+
+    oe_setup_ecall_context(&ecall_context);
+
+    run.tcs = (uint64_t)tcs;
+    run.user_handler = (uint64_t)oe_vdso_user_handler;
+    run.user_data = (uint64_t)&return_args;
+
+    while (1)
+    {
+        /* Compiler will usually handle this on exiting a function that uses
+         * AVX, but we need to avoid the AVX-SSE transition penalty here
+         * manually as part of the transition to enclave. See
+         * https://software.intel.com/content/www/us/en/develop/articles
+         * /avoiding-avx-sse-transition-penalties.html */
+        if (oe_is_avx_enabled)
+            OE_VZEROUPPER;
+
+        /* Define register bindings and initialize the registers. */
+        OE_DEFINE_REGISTER(rax, _vdso_sgx_enter_enclave);
+        OE_DEFINE_REGISTER(rbx, &run);
+        OE_DEFINE_REGISTER(rcx, ENCLU_EENTER);
+        OE_DEFINE_REGISTER(rdx, &ecall_context);
+        OE_DEFINE_REGISTER(rdi, arg1);
+        OE_DEFINE_REGISTER(rsi, arg2);
+        OE_DEFINE_REGISTER(r8, 0);
+        OE_DEFINE_REGISTER(r9, 0);
+
+        /* Save and restore MXCSR, x87 control word, RFLAGS, and non-volatile
+         * registers (excpet for RSP and RBP that are expected to be preserved)
+         * before and after the vDSO call conform the Linux x64 ABI */
+        asm volatile(
+            "stmxcsr %[mxcsr] \n\t" // Save MXCSR
+            "fstcw %[fcw] \n\t"     // Save x87 control word
+            "pushfq \n\t"           // Save RFLAGS
+            "pushq %%rbx \n\t"      // Pass 7th argument (run) via stack
+            "call *%%rax \n\t"      // Invoke _vdso_sgx_enter_enclave
```

Comment:
Hopefully we are aligning the RSP correctly for the call instruction. If not there will be a SEGV (I think).

Reply:
The original RSP should be 16-byte aligned. Two pushq (8+8) should maintain the alignment.